### DISCUSSION TITLE
GRPCRoute and Standard Channel

### DISCUSSION THREAD

**Discussion by robscott** [kubernetes-sigs/gateway-api#2844] [General]
With https://github.com/kubernetes-sigs/gateway-api/pull/2745 complete, GRPCRoute has met the criteria for graduation to standard channel in the v1.1 release. That means that it's going to be the first API to test our [updated versioning policy](https://gateway-api.sigs.k8s.io/concepts/versioning/#api-versions) that essentially skips beta.  

It likely does not make sense to carry over any notion of the v1alpha2 version of this API to standard channel, so that would mean we'd end up with the following:

| Channel | CRD API Versions | Storage Version |
| - | - | - |
| Experimental | v1alpha2, v1 | v1alpha2 |
| Standard | v1 | v1 |

All of that should work reasonably well in most cases. Here's what that means for users upgrading to v1.1 CRDs:

| Before Upgrade | After Upgrade | Result | 
| - | - | - |
| Standard | Standard | ✅ |
| Standard | Experimental | ✅ |
| Experimental | Experimental | ✅ |
| Experimental | Standard | ❌ |

As you can tell, upgrades from "experimental" to "standard" are always risky/unsafe because you're moving from an API with more things to an API with less things. In this case, it would actually be impossible to upgrade v1.0 Experimental GRPCRoute directly to v1.1 Standard GRPCRoute. That's because you'd be trying to upgrade to a CRD that had no awareness of the storage version that you were using for GRPCRoute (given that v1.1 standard likely doesn't include alpha API versions).

There are some possible workarounds here that would enable the migration path from experimental to standard. Some examples include:

1. We use v1 as the storage version for the experimental CRD - not 100% sure this is safe, will need some testing
2. Users upgrade to v1.1 experimental GRPCRoute
3. Users either manually update storage versions of their GRPCRoutes by sending empty updates to each of them, or use https://github.com/kubernetes-sigs/kube-storage-version-migrator to automate that process
4. Users upgrade to v1.1 standard GRPCRoute

This is likely a lot of thought for an upgrade path that probably won't be that common, but just wanted to get a conversation started around this because it's a fairly complex area and it will be our first time supporting this specific form of transition.

**Comment by mikemorris** [comment]
Worth considering how kube-storage-version-migrator moving in-tree for k8s 1.30 could facilitate this workflow, refs https://github.com/kubernetes/enhancements/issues/4192 and https://github.com/kubernetes/kubernetes/pull/123344
